src/Language/PureScript/CodeGen/JS.hs

```diff
@@ -65,40 +66,44 @@ moduleToJs (Module _ coms mn _ imps exps reExps foreigns decls) foreign_ =
       . (\\ (mn : C.primModules)) $ ordNub $ map snd imps
     F.traverse_ (F.traverse_ checkIntegers) optimized
     comments <- not <$> asks optionsNoComments
-    let strict = AST.StringLiteral Nothing "use strict"
-    let header = if comments && not (null coms) then AST.Comment Nothing coms strict else strict
-    let foreign' = [AST.VariableIntroduction Nothing "$foreign" foreign_ | not $ null foreigns || isNothing foreign_]
-    let moduleBody = header : foreign' ++ jsImports ++ concat optimized
+    let header = if comments && not (null coms) then AST.Comment Nothing coms else id
+    let foreign' = maybe [] (pure . AST.Import Nothing "$foreign") $ if null foreigns then Nothing else foreignInclude
+    let moduleBody = (maybe [] (uncurry (:) . first header) . uncons) $ foreign' ++ jsImports ++ concat optimized
     let foreignExps = exps `intersect` foreigns
     let standardExps = exps \\ foreignExps
     let reExps' = M.toList (M.withoutKeys reExps (S.fromList C.primModules))
-    let exps' = AST.ObjectLiteral Nothing $ map (mkString . runIdent &&& AST.Var Nothing . identToJs) standardExps
-                               ++ map (mkString . runIdent &&& foreignIdent) foreignExps
-                               ++ concatMap (reExportPairs mnLookup) reExps'
-    return $ moduleBody ++ [AST.Assignment Nothing (accessorString "exports" (AST.Var Nothing "module")) exps']
+    return $ moduleBody
+      ++ (maybeToList . exportsToJs foreignInclude $ foreignExps)
+      ++ (maybeToList . exportsToJs Nothing $ standardExps)
+      ++  mapMaybe reExportsToJs reExps'
 
   where
+  -- | Adds purity annotations to top-level values for bundlers.
+  -- The semantics here derive from treating top-level module evaluation as pure, which lets
+  -- us remove any unreferenced top-level declarations. To achieve this, we wrap any non-trivial
+  -- top-level values in an IIFE marked with a pure annotation.
+  annotatePure :: AST -> AST
+  annotatePure js@(AST.App _ (AST.Function _ _ _ _) _) = AST.Pure Nothing js
+  annotatePure js@(AST.App _ _ _) = pureIife js
+  annotatePure js@(AST.Unary _ _ _) = pureIife js
+  annotatePure js@(AST.Binary _ _ _ _) = pureIife js
+  annotatePure js@(AST.Indexer _ _ _) = pureIife js
+  annotatePure js@(AST.ObjectLiteral _ []) = js
+  annotatePure js@(AST.ObjectLiteral _ _) = pureIife js
+  annotatePure js@(AST.ArrayLiteral _ []) = js
+  annotatePure js@(AST.ArrayLiteral _ _) = pureIife js
+  annotatePure (AST.VariableIntroduction ss name (Just js)) = AST.VariableIntroduction ss name (Just (annotatePure js))
+  annotatePure (AST.Comment a b js) = AST.Comment a b (annotatePure js)
+  annotatePure js = js
+
+  pureIife :: AST -> AST
+  pureIife val = AST.Pure Nothing $ AST.App Nothing (AST.Function Nothing Nothing [] (AST.Block Nothing [AST.Return Nothing val])) []
```

Comment:
How was this arrived at? In particular, it seems a little heavy-handed to wrap every top-level non-empty object or array in an IIFE, and I think for esbuild the only things that need this annotation are function calls not in the scope of a function body. Would a more targeted approach not have worked/was it tried and found wanting/is there another bundler the working group was looking at that has different needs?

Reply:
For specific numbers, with a targeted approach, it was difficult to get esbuild to even beat purs-bundle, despite using a much lighter weight bundle encoding. It was leaving in tons of unused, seemingly trivial things, which is how I arrived at the record accessor realization. With this change, esbuild beats purs-bundle by about 30%.

Reply:
That's disappointing, but very believable. I'm fine with moving forward with this as-is, but do you by chance have a script for running those benchmarks? I'd be interested in seeing if I can improve this at some point; extra IIFEs are not as simple to strip out as comments are.

Reply:
You can follow my journey and implementations in the original PR to the working group https://github.com/working-group-purescript-es/purescript/pull/7

Reply:
As far as the scripts, they were in a branch of real-world-halogen I think. I don't remember exactly at this point. @sigma-andex might remember.

Reply:
here is the link: https://github.com/working-group-purescript-es/challenges/tree/main/bundle-challenge#purescript-halogen-template

Reply:
Thank you both!